### ISSUE TITLE
Lack of configuration file can lead to wall of output

### ISSUE THREAD

**Issue by reidmain** [realm/SwiftLint#949]
I inherited a Xcode project that already had swiftlint integrated correctly. I wrote a number of custom rules in our `.swiftlint.yml` file and they immediately popped up in Xcode. All of those custom rules were configured to raise warnings so that they didn't block developers locally but would get caught on CI where we treated errors as warnings.

What I noticed was that swiftlint wasn't getting run on our CI boxes and I couldn't (and actually still  can't) figure out why. We are using Fastlane to run our tests on CI so when I googled `Fastlane swiftlint` I immediately got kicked to https://docs.fastlane.tools/actions/#swiftlint and attempted to add the `swiftlint` action to my Fastfile.

I added:
```
swiftlint(config_file: "my_company/.swiftlint.yml",
              strict: true,
              ignore_exit_status: false)
```

and I got the unexpected output of:

```
[11:45:21]: ▸ Linting 'AppDelegate.swift' (1/370)
[11:45:21]: ▸ configuration error: 'conditional_binding_cascade' is not a valid rule identifier
[11:45:21]: ▸ Valid rule identifiers:
[11:45:21]: ▸ opening_brace
[11:45:21]: ▸ empty_count
[11:45:21]: ▸ colon
[11:45:21]: ▸ type_name
[11:45:21]: ▸ explicit_init
[11:45:21]: ▸ trailing_semicolon
[11:45:21]: ▸ nimble_operator
[11:45:21]: ▸ force_unwrapping
[11:45:21]: ▸ mark
[11:45:21]: ▸ function_body_length
[11:45:21]: ▸ variable_name
[11:45:21]: ▸ vertical_whitespace
[11:45:21]: ▸ operator_whitespace
[11:45:21]: ▸ closure_spacing
[11:45:21]: ▸ legacy_constant
[11:45:21]: ▸ line_length
[11:45:21]: ▸ return_arrow_whitespace
[11:45:21]: ▸ private_unit_test
[11:45:21]: ▸ conditional_returns_on_newline
[11:45:21]: ▸ legacy_constructor
[11:45:21]: ▸ weak_delegate
[11:45:21]: ▸ valid_docs
[11:45:21]: ▸ file_length
[11:45:21]: ▸ valid_ibinspectable
[11:45:21]: ▸ legacy_nsgeometry_functions
[11:45:21]: ▸ trailing_newline
[11:45:21]: ▸ custom_rules
[11:45:21]: ▸ overridden_super_call
[11:45:21]: ▸ comma
[11:45:21]: ▸ force_cast
[11:45:21]: ▸ private_outlet
[11:45:21]: ▸ trailing_comma
[11:45:21]: ▸ force_try
[11:45:21]: ▸ nesting
[11:45:21]: ▸ function_parameter_count
[11:45:21]: ▸ control_statement
[11:45:21]: ▸ redundant_nil_coalescing
[11:45:21]: ▸ implicit_getter
[11:45:21]: ▸ trailing_whitespace
[11:45:21]: ▸ closing_brace
[11:45:21]: ▸ statement_position
[11:45:21]: ▸ type_body_length
[11:45:21]: ▸ todo
[11:45:21]: ▸ cyclomatic_complexity
[11:45:21]: ▸ missing_docs
[11:45:21]: ▸ switch_case_on_newline
[11:45:21]: ▸ syntactic_sugar
[11:45:21]: ▸ leading_whitespace
[11:45:21]: ▸ legacy_cggeometry_functions
[11:45:21]: ▸ Linting 'BankViewController.swift' (2/370)
[11:45:21]: ▸ configuration error: 'conditional_binding_cascade' is not a valid rule identifier
[11:45:21]: ▸ Valid rule identifiers:
[11:45:21]: ▸ opening_brace
[11:45:21]: ▸ empty_count
[11:45:21]: ▸ colon
[11:45:21]: ▸ type_name
[11:45:21]: ▸ explicit_init
[11:45:21]: ▸ trailing_semicolon
[11:45:21]: ▸ nimble_operator
[11:45:21]: ▸ force_unwrapping
[11:45:21]: ▸ mark
[11:45:21]: ▸ function_body_length
[11:45:21]: ▸ variable_name
[11:45:21]: ▸ vertical_whitespace
[11:45:21]: ▸ operator_whitespace
[11:45:21]: ▸ closure_spacing
[11:45:21]: ▸ legacy_constant
[11:45:21]: ▸ line_length
[11:45:21]: ▸ return_arrow_whitespace
[11:45:21]: ▸ private_unit_test
[11:45:21]: ▸ conditional_returns_on_newline
[11:45:21]: ▸ legacy_constructor
[11:45:21]: ▸ weak_delegate
[11:45:21]: ▸ valid_docs
[11:45:21]: ▸ file_length
[11:45:21]: ▸ valid_ibinspectable
[11:45:21]: ▸ legacy_nsgeometry_functions
[11:45:21]: ▸ trailing_newline
[11:45:21]: ▸ custom_rules
[11:45:21]: ▸ overridden_super_call
[11:45:21]: ▸ comma
[11:45:21]: ▸ force_cast
[11:45:21]: ▸ private_outlet
[11:45:21]: ▸ trailing_comma
[11:45:21]: ▸ force_try
[11:45:21]: ▸ nesting
[11:45:21]: ▸ function_parameter_count
[11:45:21]: ▸ control_statement
[11:45:21]: ▸ redundant_nil_coalescing
[11:45:21]: ▸ implicit_getter
[11:45:21]: ▸ trailing_whitespace
[11:45:21]: ▸ closing_brace
[11:45:21]: ▸ statement_position
[11:45:21]: ▸ type_body_length
[11:45:21]: ▸ todo
[11:45:21]: ▸ cyclomatic_complexity
[11:45:21]: ▸ missing_docs
[11:45:21]: ▸ switch_case_on_newline
[11:45:21]: ▸ syntactic_sugar
[11:45:21]: ▸ leading_whitespace
[11:45:21]: ▸ legacy_cggeometry_functions
```

For some reason the rules that were being used were being outputted for every file that was linted. With 370 files my console immediately became an unreadable mess.

The command that Fastlane was running was `swiftlint lint --strict --config my_company/.swiftlint.yml` and when I ran the command in the terminal I got the same output as above.

However when I `cd my_company` and ran `swiftlint lint --strict --config .swiftlint.yml` I got what you would expect. The rules only printed out once and then each file being linted was listed.

```
Loading configuration from '.swiftlint.yml'
configuration error: 'conditional_binding_cascade' is not a valid rule identifier
Valid rule identifiers:
opening_brace
empty_count
colon
type_name
explicit_init
trailing_semicolon
nimble_operator
force_unwrapping
mark
function_body_length
variable_name
vertical_whitespace
operator_whitespace
closure_spacing
legacy_constant
line_length
return_arrow_whitespace
private_unit_test
conditional_returns_on_newline
legacy_constructor
weak_delegate
valid_docs
file_length
valid_ibinspectable
legacy_nsgeometry_functions
trailing_newline
custom_rules
overridden_super_call
comma
force_cast
private_outlet
trailing_comma
force_try
nesting
function_parameter_count
control_statement
redundant_nil_coalescing
implicit_getter
trailing_whitespace
closing_brace
statement_position
type_body_length
todo
cyclomatic_complexity
missing_docs
switch_case_on_newline
syntactic_sugar
leading_whitespace
legacy_cggeometry_functions
Linting Swift files in current working directory
Linting 'AppDelegate.swift' (1/298)
Linting 'BankViewController.swift' (2/298)
Linting 'BankSearchViewController.swift' (3/298)
```

I then went back to the original directory, ran just `swiftlint` and got the same issue as before where every file that was linted outputted the rules being used.

It looks like if swiftlint doesn't have a configuration file specified or if one does not exist in the same directory that swiftlint is being run from there is some issue that causes the rules to be outputted for every file which gunks up the console.

**Comment by jpsim** [comment]
Thanks for the thorough report!

What happens when you run the same command Fastlane does?

```shell
$ swiftlint lint --strict --config "my_company/.swiftlint.yml"
```

I get the feeling that [Configuration. configurationForPath(_:)](https://github.com/realm/SwiftLint/blob/0.13.2/Source/SwiftLintFramework/Models/Configuration.swift#L220) gets called for each file because the config file isn't in root path...

If that's the case, caching configurations per path might be the quickest way to fix this.

**Comment by jpsim** [comment]
Hopefully #951 fixes this. I'd appreciate if you could confirm this fixes the issue in your case @reidmain!

**Comment by reidmain** [comment]
@jpsim sure I will look at pulling master of the repo and building swiftlint. I've only ever used it via homebrew up until this point.

**Comment by jpsim** [comment]
Actually it's not _on_ `master`, since #951 hasn't been merged yet. So you'd need to build from the [`jp-fix-multi-config-logs`](https://github.com/realm/SwiftLint/tree/jp-fix-multi-config-logs) branch.

**Comment by reidmain** [comment]
@jpsim yeah I caught that after I actually did pull master :p I updated the comment and am about to run `make install` right now.

**Comment by reidmain** [comment]
So if I run `swiftlint lint --strict --config "my_company/.swiftlint.yml"` I now get:

```
Loading configuration from 'affirm/.swiftlint.yml'
configuration error: 'conditional_binding_cascade' is not a valid rule identifier
Valid rule identifiers:
empty_count
trailing_semicolon
type_name
statement_position
todo
legacy_constant
force_cast
nimble_operator
number_separator
comma
implicit_getter
legacy_cggeometry_functions
force_unwrapping
leading_whitespace
cyclomatic_complexity
function_body_length
control_statement
type_body_length
operator_whitespace
closure_spacing
vertical_whitespace
valid_docs
trailing_whitespace
mark
legacy_nsgeometry_functions
closing_brace
colon
nesting
switch_case_on_newline
conditional_returns_on_newline
trailing_newline
missing_docs
variable_name
redundant_nil_coalescing
private_unit_test
return_arrow_whitespace
overridden_super_call
private_outlet
explicit_init
legacy_constructor
custom_rules
function_parameter_count
syntactic_sugar
trailing_comma
valid_ibinspectable
opening_brace
line_length
weak_delegate
force_try
file_length
Linting Swift files in current working directory
Linting 'AppDelegate.swift' (1/370)
fatal error: Could not read configuration file at path '/Users/reidmain/Development/Affirm/ios/Affirm': Error Domain=NSCocoaErrorDomain Code=257 "The file “affirm” couldn’t be opened because you don’t have permission to view it." UserInfo={NSFilePath=/Users/reidmain/Development/Affirm/ios/Affirm, NSUnderlyingError=0x7fb104527d80 {Error Domain=NSPOSIXErrorDomain Code=13 "Permission denied"}}: file /Users/reidmain/Development/GitHub/SwiftLint/Source/SwiftLintFramework/Models/Configuration.swift, line 164
Illegal instruction: 4
```

I was trying to anonymize my company name in the previous logs but I feel like I shouldn't do that in this one because `/Users/reidmain/Development/Affirm/ios/Affirm` is a really weird path. It physically doesn't exist. `/Users/reidmain/Development/Affirm/ios/affirm` does however which is where the config file lives.

**Comment by reidmain** [comment]
Oh but I should mention that if I run `swiftlint` at the same level as the configuration file it still works.

Also this may be useful. If I just run `swiftlint` in the directory that fastlane would run it in I get

```
swiftlint
Linting Swift files in current working directory
Linting 'AppDelegate.swift' (1/370)
fatal error: Could not read configuration file at path '/Users/reidmain/Development/Affirm/ios/affirm': Error Domain=NSCocoaErrorDomain Code=257 "The file “affirm” couldn’t be opened because you don’t have permission to view it." UserInfo={NSFilePath=/Users/reidmain/Development/Affirm/ios/affirm, NSUnderlyingError=0x7fe6ec72f550 {Error Domain=NSPOSIXErrorDomain Code=13 "Permission denied"}}: file /Users/reidmain/Development/GitHub/SwiftLint/Source/SwiftLintFramework/Models/Configuration.swift, line 164
Illegal instruction: 4
```

**Comment by jpsim** [comment]
And this is with SwiftLint built from source from the `jp-fix-multi-config-logs` branch?

**Comment by reidmain** [comment]
@jpsim that is correct. I pulled down that branch and ran `make install`. I assume it overrode whatever I had installed through homebrew.

**Comment by jpsim** [comment]
yes, that should override homebrew

**Comment by jpsim** [comment]
I might not be able to get back to this for a little bit. Could you ping me here if you haven't heard back in a few days?

**Comment by reidmain** [comment]
@jpsim sure there is no rush for this. Just one of those CI cleanup tasks I spend time on every so often.

**Comment by reidmain** [comment]
ping @jpsim

**Comment by reidmain** [comment]
🏓

**Comment by jpsim** [comment]
lol, I had a good chuckle with your humorous pings & pongs. Finally revived #951.

**Comment by jpsim** [comment]
Fixed in #951 (but still pending confirmation).

**Comment by reidmain** [comment]
Looks like this was actually fixed somewhere between 0.13.1 and 0.18.1

I can no longer replicate this issue on 0.18.1
